vpr/src/route/connection_box_lookahead_map.h

```diff
@@ -4,14 +4,116 @@
 #include <vector>
 #include "physical_types.h"
 #include "router_lookahead.h"
+#include "router_lookahead_map_utils.h"
+#include "connection_box.h"
+#include "vtr_geometry.h"
 
+// Keys in the RoutingCosts map
+struct RoutingCostKey {
+    // segment type index
+    int seg_index;
+
+    // type of the destination connection box
+    ConnectionBoxId box_id;
+
+    // offset of the destination connection box from the starting segment
+    vtr::Point<int> delta;
+
+    RoutingCostKey()
+        : seg_index(-1)
+        , delta(0, 0) {}
+    RoutingCostKey(int seg_index_arg, ConnectionBoxId box_id_arg, vtr::Point<int> delta_arg)
+        : seg_index(seg_index_arg)
+        , box_id(box_id_arg)
+        , delta(delta_arg) {}
+
+    bool operator==(const RoutingCostKey& other) const {
+        return seg_index == other.seg_index && box_id == other.box_id && delta == other.delta;
+    }
+};
+
+// compressed version of RoutingCostKey
+// TODO add bounds checks
+struct CompressedRoutingCostKey {
+    uint32_t data;
+
+    CompressedRoutingCostKey() {
+        data = -1;
+    }
+    CompressedRoutingCostKey(const RoutingCostKey& key) {
+        data = key.seg_index & 0xff;
+        data <<= 8;
+        data |= size_t(key.box_id) & 0xff;
+        data <<= 8;
+        data |= key.delta.x() & 0xff;
+        data <<= 8;
+        data |= key.delta.y() & 0xff;
+    }
+
+    bool operator==(CompressedRoutingCostKey other) const {
+        return data == other.data;
+    }
+};
+
+// Data in the RoutingCosts map
+struct RoutingCost {
+    // source and destination node indices
+    int from_node, to_node;
+
+    // cost entry for the route
+    Cost_Entry cost_entry;
+};
+
+// hash implementation for RoutingCostKey
+struct HashRoutingCostKey {
+    std::size_t operator()(RoutingCostKey const& key) const noexcept {
+        uint64_t data;
```

Comment:
Why use this hash over `vtr::hash_combine`?

Reply:
It packs everything together into something that can be hashed once, so it should be faster. I haven't verified that yet. It was inspired by `CompressedRoutingCostKey`.

Reply:
Unless you have data showing the using `std::hash` + `vtr::hash_combine` is significantly slow than this, please leave this as `std::hash` + `vtr::hash_combine`.

Reply:
I've switched it to `vtr::hash_combine`.

Reply:
I've changed it to use `vtr::hash_combine`.